### PR TITLE
Page size preference

### PR DESCRIPTION
**Pull request by milleymilli**
This is a:

- [x] ✨ **New feature** - new behaviour has been implemented
- [ ] 🐛 **Bug fix** - existing behaviour has been made to behave
- [ ] ♻️ **Refactor** - the behaviour has not changed, just the implementation
- [ ] ✅ **Test backfill** - tests for existing behaviour were added but the behaviour itself hasn't changed
- [ ] ⚙️ **Chore** - maintenance task, behaviour and implementation haven't changed

### Description

- **Purpose** -
   - **Page Size Preference**: 
      Users will have the ability to select their preferred page size using a drop down UI component.
      This selected page size preference will be stored in the local storage of the user's browser for future visits to the application.
    - **Local Storage Management**:
    The application will utilize the browser's local storage to save and retrieve the user's chosen page size preference.
    This enables persistent storage of the selected preference, maintaining it even if the user refreshes the page or returns to the application later.
   - **Enhanced User Experience**:
    Overall, merging this new code will enhance the user experience by providing an interactive feature that allows users to customize their browsing experience by selecting their preferred page size for resource display.

- **How to check** -
   - **Inspect Local Storage Changes**:
    Use browser developer tools to inspect the local storage for the application.
    Look for the presence of stored data related to the page size preference and    confirm that it persists across sessions or page reloads.
### Author checklist

- [x] I have written a title that reflects the relevant ticket
- [x] I have written a description that says what the PR does and how to validate it
- [ ] I have linked to the project board ticket (and any related PRs/issues) in the Links section
- [ ] I have added a link to this PR to the ticket
- [ ] I have made the PR to `main` from a branch named `<category>/<name>`, e.g. `feature/edit-spaceships` or `bugfix/restore-oxygen`
- [x] I have manually tested that the app still works correctly
- [ ] I have requested reviewers here and in my team chat channel

- [ ] I have spoken with my PM or TL about any parts of this task that may have become out-of-scope, or any additional improvements that I now realise may benefit my project
- [x] I have added tests, or new tests were not required
- [x] I have updated any documentation (e.g. diagrams, schemas), or documentation updates were not required

### CODE REVIEW

**Review by milleymilli — COMMENTED** [review]
This PR introduces a mechanism to manage the user's preferred page size for better customization of the application experience. Here's a breakdown of the changes:

1. **Fetching Initial Page Size:**
   - Reads the user's preferred page size from local storage.
   - If available, attempts to parse the stored value as JSON.
   - In case of parsing error, defaults to "default" and throws an error for debugging purposes.

2. **useState and useEffect Integration:**
   - Utilizes the `useState` hook to manage the current page size, initializing it with the parsed or default value.
   - `useEffect` to synchronize the selected page size with API requests.

3. **LocalStorage Interaction:**
   - Implements a `savePageSizePreference` function to store the selected page size in local storage.

These changes enhance the user experience by allowing them to customize their preferred page size. The error handling in the parsing process aims to provide better debugging insights if issues arise.

Please review and provide feedback. Thank you!

**Review by textbook — CHANGES_REQUESTED** [review]
Please complete the template in the PR description to enable review.

**Comment by milleymilli** [comment]
> Please complete the template in the PR description to enable review.

I have completed it.

File: client/src/pages/Home/index.js
Comment (textbook) [inline client/src/pages/Home/index.js:1]: ❓ Why was this import added? It's not explicitly used anywhere, and [the "new" transform][1] means this is no longer required in JSX files generally.

[1]: https://legacy.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html

File: client/src/pages/Home/index.js
Comment (textbook) [inline client/src/pages/Home/index.js:7]: ❓ Why switch from vanilla to arrow function here? I don't think any of the other components in this project are arrow functions.

File: client/src/pages/Home/index.js
Comment (textbook) [inline client/src/pages/Home/index.js:10]: ❓ Why was this change introduced? Had you experienced some issue with the previous implementation?

File: client/src/pages/Home/index.js
Comment (textbook) [inline client/src/pages/Home/index.js:37]: ❌  Throwing an error in this way is generally not helpful - if you want to created a new error "caused by" the previous one it should be:

```suggestion
				throw new Error("Error fetching published resources:", { cause: error });
```

[MDN][1]

[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/Error#rethrowing_an_error_with_a_cause

But if you can't really deal with the error or add any useful information, it's best to let it keep going to wherever it ends up.

File: client/src/pages/Home/index.js
Comment (textbook) [inline client/src/pages/Home/index.js:33]: ❌ What does adding `pageSize` to the arguments here achieve? If you look at the service implementation:

https://github.com/CodeYourFuture/tech-products-demo/blob/40bf4ae5c387cc19828e1c2747f63e41fa6f5369/client/src/services/resourceService.js#L19

It destructures `page` and `perPage` out, but anything else is ignored. You could also see by looking at the Network tab in your developer tools that the request had no additional parameters. Even if that _wasn't_ true, and it passed `pageSize` along to the request, [the `GET /resources` endpoint](https://tech-products-demo-d1182e30f280.herokuapp.com/docs/#/Resources/get_resources) doesn't document a `pageSize` query parameter and would at best ignore it (search/query parameters tend to be optional, although some API implementations will validate and reject 4xx on unexpected parameters/values).

File: client/src/pages/Home/index.js
Comment (textbook) [inline client/src/pages/Home/index.js:56]: ❓ What are these values supposed to represent? Assuming they follow what's already on the page, i.e.:

- Default: 20
- Small: 10
- Large: 50

it would be helpful to include those values, and this is an odd order to put them in.

File: client/src/pages/Home/index.js
Comment (textbook) [inline client/src/pages/Home/index.js:43]: 💡  I'd generally discourage comments that simply reiterate what the code _does_. In this case you could have clarified this by _extracting a component_, rather than jamming everything directly into the Home page, then its name would have conveyed its usage e.g. `<SelectPageSize {...props} />`.

File: client/src/pages/Home/index.js
Comment (textbook) [inline client/src/pages/Home/index.js:28]: 💡 This is a lot of code in the middle of the home page to handle something relatively minor (and that might also be needed _elsewhere_). Don't just jam everything into one file, think about how you can separate out unrelated concerns to make your code easier to read (and test). For example, what if you extracted a _service_ to handle local storage, or maybe a _hook_:
```suggestion
const [pageSize, setPageSize] = useLocalStorage("pageSizePreference", "default");
```
Also note some of my other suggestions (error handling, comments\*) apply here too.

\* _if you're going to put comments on functions, use [JSDoc](https://jsdoc.app/) so they're structured and useful_

**Review by textbook — CHANGES_REQUESTED** [review]
**PR**:

❌ Still doesn't link to the issue: #17
❌ Checklist is incomplete/incorrect (e.g. tests are checked, but there aren't any)

I was also confused by the "How to check", as it didn't seem to refer to any actual functionality, but...

---

**Functionality**:

❌ **Doesn't actually do anything.** Yes it updates local storage, and sets the select accordingly across sessions and page re-loads. But what _use_ is that? It doesn't actually change how many items are in the page.
❓ What about other lists in the UI? Admins can see a list of drafts, should that follow this preference too?
❓ How would you approach this if the requirement was to remember a user's preference across browsers/computers?

---

**UI**:

Screenshot 2023-11-14 at 16 34 39

❌ Select isn't consistently styled, doesn't match the other one (10, 20, 50) on the same page
❌ Label isn't consistent either ("Title Case" instead of "Sentence case", includes a colon, no whitespace before control) 
❓ Is this the right place for it? It's a long way from the other page-related controls, especially if there are any resources in the list.

---

More detailed comments on the implementation details below. In general:

- It's not well factored, look for opportunities to split out separate parts rather than add eveything into one file; and
- It's completely untested.

**Key**:

- ❌ Problem, needs fixing before approval
- ❓ Question, needs answering before approval
- 💡 Suggestion, can be taken or left

**Comment by textbook** [comment]
🍞 Closing as stale (more than one month without updates)

**Pull request closed without merging**
